File: Leetcode_with_py/155MinStack.py
class MinStack(object):

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.val = []
        self.index = -1
        
    def push(self, x):
        """
        :type x: int
        :rtype: None
        """
        self.val.append(x)
        self.index += 1

    def pop(self):
        """
        :rtype: None
        """
        self.val.pop()
        self.index -= 1

    def top(self):
        """
        :rtype: int
        """
        return self.val[self.index]

    def getMin(self):
        """
        :rtype: int
        """
        return min(self.val)
        
# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(x)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: Leetcode_with_Cpp/155MinStack.cpp
class MinStack {
private:
    stack<int> data_stack;
    stack<int> min_stack;
public:
    /** initialize your data structure here. */
    MinStack() {}
    
    void push(int x) {
        if (min_stack.size() == 0 || min_stack.top() >= x) {
            min_stack.push(x);
        }
        data_stack.push(x);
    }
    
    void pop() {
        if (data_stack.top() == min_stack.top()) {
            min_stack.pop();
        }
        data_stack.pop();
    }
    
    int top() {
        return data_stack.top();
    }
    
    int getMin() {
        return min_stack.top();
    }
};
/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(x);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */




class MinStack {
public:
    /** initialize your data structure here. */
    struct Node {
        int val;
        int min;
        Node *next;
        Node(int _val, Node *_next = NULL) : val(_val), min(val), next(_next) {}
    };
    Node *head;

    MinStack() {
        head = new Node(0);
    }
  
    void push(int x) {
        Node *tmp = new Node(x, head->next);
        if (head->next && x > head->next->min) {  // tmp->min 记录该节点之前所有节点最小值； 新建节点加到 head->next 以便 pop;
			tmp->min = head->next->min;
		}
		head->next = tmp;
    }
    
    void pop() {
        Node *tmp = head->next;
		if (tmp) {
			head->next = tmp->next;
			delete tmp;
		}
    }
    
    int top() {
        int top = 0;
		if (head->next) {
			top = head->next->val;
		}
		return top;
    }
    
    int getMin() {
        int min = 0;
		if (head->next) {
			min = head->next->min;
		}
		return min;
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(x);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */